### PR TITLE
Add the ckit command

### LINKED ISSUE #226: [Suggestion] `/ckits`
So this is just like `/cgive`, its only accessible in creative.
So sometimes, for example when we making redstone machine, the saved toolbar just not enough (only 9 items). So this command can delete, load, and save inventory

Syntax :
`/cinventory (load|save|delete) <file name (without .txt)>`
and then maybe store the inventory in `.minecraft/config/clientcommands/inventory/` with txt format?
So inside `.minecraft/config/clientcommands/inventory/<file name>.txt` will be
```
<armor.helmet>
<armor.chestplate>
<armor.leggings>
<armor.boots>

<weapon.offhand>

<inventory.0>
....
<inventory.36>

<hotbar.0>
....
<hotbar.8>
```

### PR DESCRIPTION
**Pull request by xpple**
Closes #226. I sticked to Minecraft's style of storing `.dat` files as much as possible. As suggested by Earthcomputer, all the kits are stored in one file.

### CODE REVIEW

File: src/main/java/net/earthcomputer/clientcommands/command/KitCommand.java [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+import java.io.File;
+import java.nio.file.Path;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import static net.earthcomputer.clientcommands.command.ClientCommandManager.addClientSideCommand;
+import static net.earthcomputer.clientcommands.command.ClientCommandManager.sendFeedback;
+import static net.minecraft.server.command.CommandManager.argument;
+import static net.minecraft.server.command.CommandManager.literal;
```
Comment (Earthcomputer) [inline src/main/java/net/earthcomputer/clientcommands/command/KitCommand.java]: import statics should normally use wildcard imports

File: src/main/java/net/earthcomputer/clientcommands/command/KitCommand.java [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+
+public class KitCommand {
+
+    private static final SimpleCommandExceptionType SAVE_FILE_EXCEPTION = new SimpleCommandExceptionType(new TranslatableText("commands.ckit.save.failed"));
```
Comment (Earthcomputer) [inline src/main/java/net/earthcomputer/clientcommands/command/KitCommand.java]:
```suggestion
    private static final SimpleCommandExceptionType SAVE_FAILED_EXCEPTION = new SimpleCommandExceptionType(new TranslatableText("commands.ckit.save.failed"));
```

File: src/main/java/net/earthcomputer/clientcommands/command/KitCommand.java [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+    private static final SimpleCommandExceptionType LOAD_FILE_EXCEPTION = new SimpleCommandExceptionType(new TranslatableText("commands.ckit.load.failed"));
```
Comment (Earthcomputer) [inline src/main/java/net/earthcomputer/clientcommands/command/KitCommand.java]:
```suggestion
    private static final SimpleCommandExceptionType LOAD_FAILED_EXCEPTION = new SimpleCommandExceptionType(new TranslatableText("commands.ckit.load.failed"));
```

File: src/main/java/net/earthcomputer/clientcommands/command/KitCommand.java [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+
+    private static final SimpleCommandExceptionType NOT_CREATIVE_EXCEPTION = new SimpleCommandExceptionType(new TranslatableText("commands.ckit.kit.notCreative"));
+    private static final DynamicCommandExceptionType NOT_FOUND_EXCEPTION = new DynamicCommandExceptionType(arg -> new TranslatableText("commands.ckit.kit.notFound", arg));
+
+    private static final Path path = FabricLoader.getInstance().getConfigDir().resolve("clientcommands");
```
Comment (Earthcomputer) [inline src/main/java/net/earthcomputer/clientcommands/command/KitCommand.java]: This could have a more descriptive name than `path`

File: src/main/java/net/earthcomputer/clientcommands/command/KitCommand.java [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+
+        client.player.playerScreenHandler.sendContentUpdates();
+        sendFeedback("commands.ckit.kit.success", name);
+        return 1;
+    }
+
+    private static int kits(ServerCommandSource source) throws CommandSyntaxException {
+        loadFile();
+
+        sendFeedback("Available kits: " + String.join(", ", kits.keySet()));
```
Comment (Earthcomputer) [inline src/main/java/net/earthcomputer/clientcommands/command/KitCommand.java]: String.join is a Java 9+ method

File: src/main/java/net/earthcomputer/clientcommands/command/KitCommand.java [hunk 6]
```diff
@@ -0,0 +1,182 @@
...
+            CompoundTag compoundTag = new CompoundTag();
+
+            for (Map.Entry<String, PlayerInventory> entry : kits.entrySet()) {
+                compoundTag.put(entry.getKey(), entry.getValue().serialize(new ListTag()));
+            }
+
+            File file = File.createTempFile("kits", ".dat", path.toFile());
+            NbtIo.write(compoundTag, file);
+            File file2 = new File(path.toFile(), "kits.dat_old");
+            File file3 = new File(path.toFile(), "kits.dat");
```
Comment (Earthcomputer) [inline src/main/java/net/earthcomputer/clientcommands/command/KitCommand.java]: `file`, `file1` and `file2` could have more descriptive names. Like the idea of creating a backup though :+1:

File: src/main/java/net/earthcomputer/clientcommands/command/KitCommand.java [hunk 7]
```diff
@@ -0,0 +1,182 @@
...
+            compoundTag.getKeys().forEach(key -> {
+                PlayerInventory inventory = new PlayerInventory(client.player);
+                inventory.deserialize(compoundTag.getList(key, 10));
+                kits.put(key, inventory);
+            });
+        } catch (Exception e) {
+            throw LOAD_FILE_EXCEPTION.create();
+        }
+    }
+}
```
Comment (Earthcomputer) [inline src/main/java/net/earthcomputer/clientcommands/command/KitCommand.java]: Newline at the end of the file

File: src/main/java/net/earthcomputer/clientcommands/command/KitCommand.java [hunk 8]
```diff
@@ -0,0 +1,182 @@
...
+        return 1;
+    }
+
+    private static void saveFile() throws CommandSyntaxException {
+        try {
+            CompoundTag compoundTag = new CompoundTag();
+
+            for (Map.Entry<String, PlayerInventory> entry : kits.entrySet()) {
+                compoundTag.put(entry.getKey(), entry.getValue().serialize(new ListTag()));
+            }
```
Comment (Earthcomputer) [inline src/main/java/net/earthcomputer/clientcommands/command/KitCommand.java]: this could be easily replaced by `kits.forEach`

File: src/main/java/net/earthcomputer/clientcommands/command/KitCommand.java [hunk 9]
```diff
@@ -0,0 +1,182 @@
...
+        loadFile();
+
+        if (kits.put(name, client.player.inventory) == null) {
+            saveFile();
+            sendFeedback("commands.ckit.create.success", name);
+        } else {
+            saveFile();
+            sendFeedback("commands.ckit.create.alreadyExists", name);
+        }
+        return 1;
```
Comment (Earthcomputer) [inline src/main/java/net/earthcomputer/clientcommands/command/KitCommand.java]: The convention in clientcommands is to return 0 unless there is a more sensible value to return.

File: src/main/java/net/earthcomputer/clientcommands/command/KitCommand.java [hunk 10]
```diff
@@ -0,0 +1,182 @@
...
+    private static int create(ServerCommandSource source, String name) throws CommandSyntaxException {
+        loadFile();
+
+        if (kits.put(name, client.player.inventory) == null) {
+            saveFile();
+            sendFeedback("commands.ckit.create.success", name);
+        } else {
+            saveFile();
+            sendFeedback("commands.ckit.create.alreadyExists", name);
+        }
```
Comment (Earthcomputer) [inline src/main/java/net/earthcomputer/clientcommands/command/KitCommand.java]: The main point of having the split between edit and create is to protect against people accidentally overwriting kits. So this function should throw an error if the kit already exists.

File: src/main/java/net/earthcomputer/clientcommands/command/KitCommand.java [hunk 11]
```diff
@@ -0,0 +1,182 @@
...
+        sendFeedback("commands.ckit.delete.success", name);
+        return 1;
+    }
+
+    private static int edit(ServerCommandSource source, String name) throws CommandSyntaxException {
+        loadFile();
+
+        if (kits.put(name, client.player.inventory) == null) {
+            saveFile();
+            sendFeedback("commands.ckit.edit.notFoundSoCreatedNew", name);
```
Comment (Earthcomputer) [inline src/main/java/net/earthcomputer/clientcommands/command/KitCommand.java]: Same as above, this function should throw an error if the kit does not exist.

File: src/main/java/net/earthcomputer/clientcommands/command/KitCommand.java [hunk 12]
```diff
@@ -0,0 +1,182 @@
...
+                                .executes(ctx -> create(ctx.getSource(), StringArgumentType.getString(ctx, "name")))))
+                .then(literal("delete")
+                        .then(argument("name", StringArgumentType.string())
+                                .suggests((ctx, builder) -> CommandSource.suggestMatching(kits.keySet(), builder))
+                                .executes(ctx -> delete(ctx.getSource(), StringArgumentType.getString(ctx, "name")))))
+                .then(literal("edit")
+                        .then(argument("name", StringArgumentType.string())
+                                .suggests((ctx, builder) -> CommandSource.suggestMatching(kits.keySet(), builder))
+                                .executes(ctx -> edit(ctx.getSource(), StringArgumentType.getString(ctx, "name")))))
+                .then(literal("kit")
```
Comment (Earthcomputer) [inline src/main/java/net/earthcomputer/clientcommands/command/KitCommand.java]: I think this subcommand should be called "load" or something similar.

File: src/main/java/net/earthcomputer/clientcommands/command/KitCommand.java [hunk 13]
```diff
@@ -0,0 +1,182 @@
...
+                        .then(argument("name", StringArgumentType.string())
+                                .suggests((ctx, builder) -> CommandSource.suggestMatching(kits.keySet(), builder))
+                                .then(literal("--override")
+                                        .executes(ctx -> kit(ctx.getSource(), StringArgumentType.getString(ctx, "name"), true)))
+                                .executes(ctx -> kit(ctx.getSource(), StringArgumentType.getString(ctx, "name"), false))))
+                .then(literal("kits")
```
Comment (Earthcomputer) [inline src/main/java/net/earthcomputer/clientcommands/command/KitCommand.java]: I think this subcommand should be called "list".

File: src/main/java/net/earthcomputer/clientcommands/command/KitCommand.java [hunk 14]
```diff
@@ -0,0 +1,182 @@
...
+
+    private static void loadFile() throws CommandSyntaxException {
+        try {
+            kits.clear();
+            CompoundTag compoundTag = NbtIo.read(new File(path.toFile(), "kits.dat"));
+            if (compoundTag == null) return;
+
+            compoundTag.getKeys().forEach(key -> {
+                PlayerInventory inventory = new PlayerInventory(client.player);
+                inventory.deserialize(compoundTag.getList(key, 10));
```
Comment (Earthcomputer) [inline src/main/java/net/earthcomputer/clientcommands/command/KitCommand.java]: I think Fabric API has an NbtTypes class which contains the named constants instead of this magic literal "10".

File: src/main/java/net/earthcomputer/clientcommands/command/KitCommand.java [hunk 15]
```diff
@@ -0,0 +1,182 @@
...
+    private static final SimpleCommandExceptionType LOAD_FILE_EXCEPTION = new SimpleCommandExceptionType(new TranslatableText("commands.ckit.load.failed"));
+
+    private static final SimpleCommandExceptionType NOT_CREATIVE_EXCEPTION = new SimpleCommandExceptionType(new TranslatableText("commands.ckit.kit.notCreative"));
+    private static final DynamicCommandExceptionType NOT_FOUND_EXCEPTION = new DynamicCommandExceptionType(arg -> new TranslatableText("commands.ckit.kit.notFound", arg));
+
+    private static final Path path = FabricLoader.getInstance().getConfigDir().resolve("clientcommands");
+
+    private static final MinecraftClient client = MinecraftClient.getInstance();
+
+    private static final Map<String, PlayerInventory> kits = new HashMap<>();
```
Comment (Earthcomputer) [inline src/main/java/net/earthcomputer/clientcommands/command/KitCommand.java]: I think `PlayerInventory` stores a reference to the player, which stores a reference to the world. So when switching worlds you are going to get a memory leak because you're retaining a reference to the old world. Instead, you should store the `CompoundTag`s in this map.

File: src/main/java/net/earthcomputer/clientcommands/command/KitCommand.java [hunk 16]
```diff
@@ -0,0 +1,182 @@
...
+            sendFeedback("commands.ckit.edit.notFoundSoCreatedNew", name);
+        } else {
+            saveFile();
+            sendFeedback("commands.ckit.edit.success", name);
+        }
+        return 1;
+    }
+
+    private static int kit(ServerCommandSource source, String name, boolean override) throws CommandSyntaxException {
+        if (!client.player.isCreative() || !client.player.abilities.creativeMode) throw NOT_CREATIVE_EXCEPTION.create();
```
Comment (Earthcomputer) [inline src/main/java/net/earthcomputer/clientcommands/command/KitCommand.java]: New code in clientcommands should use braces for all if statements. There are a few of them in this PR.

**Review by Earthcomputer — CHANGES_REQUESTED** [review]
Please address these review comments. (Remember not to force push from this point, as force pushing makes it harder for me to see what you changed).

**Comment by xpple** [comment]
Fuck I left some debugging in, will remove in the upcoming commit

File: src/main/java/net/earthcomputer/clientcommands/command/KitCommand.java [hunk 17]
```diff
@@ -0,0 +1,191 @@
...
+                                        .executes(ctx -> load(ctx.getSource(), StringArgumentType.getString(ctx, "name"), true)))
+                                .executes(ctx -> load(ctx.getSource(), StringArgumentType.getString(ctx, "name"), false))))
+                .then(literal("list")
+                        .executes(ctx -> list(ctx.getSource()))));
+    }
+
+    private static int create(ServerCommandSource source, String name) throws CommandSyntaxException {
+        loadFile();
+
+        if (kits.get(name) != null) {
```
Comment (Earthcomputer) [inline src/main/java/net/earthcomputer/clientcommands/command/KitCommand.java]:
```suggestion
        if (kits.containsKey(name)) {
```

File: src/main/java/net/earthcomputer/clientcommands/command/KitCommand.java [hunk 18]
```diff
@@ -0,0 +1,191 @@
...
+        }
+        saveFile();
+        sendFeedback("commands.ckit.delete.success", name);
+        return 0;
+    }
+
+    private static int edit(ServerCommandSource source, String name) throws CommandSyntaxException {
+        loadFile();
+
+        if (kits.get(name) == null) {
```
Comment (Earthcomputer) [inline src/main/java/net/earthcomputer/clientcommands/command/KitCommand.java]:
```suggestion
        if (!kits.containsKey(name)) {
```

File: src/main/java/net/earthcomputer/clientcommands/command/KitCommand.java [hunk 19]
```diff
@@ -0,0 +1,191 @@
...
+            throw NOT_FOUND_EXCEPTION.create(name);
+        }
+        kits.put(name, client.player.inventory.serialize(new ListTag()));
+        saveFile();
+        sendFeedback("commands.ckit.edit.success", name);
+        return 0;
+    }
+
+    private static int load(ServerCommandSource source, String name, boolean override) throws CommandSyntaxException {
+        if (!client.player.isCreative() || !client.player.abilities.creativeMode) {
```
Comment (Earthcomputer) [inline src/main/java/net/earthcomputer/clientcommands/command/KitCommand.java]: Surely you don't need both of these conditions?

File: src/main/java/net/earthcomputer/clientcommands/command/KitCommand.java [hunk 20]
```diff
@@ -0,0 +1,191 @@
...
+            throw NOT_FOUND_EXCEPTION.create(name);
+        }
+
+        PlayerInventory tempInv = new PlayerInventory(client.player);
+        tempInv.deserialize(kit);
+        List<Slot> slots = client.player.playerScreenHandler.slots;
+        for (int i = 0; i < slots.size(); i++) {
+            if (slots.get(i).inventory == client.player.inventory) {
+                ItemStack itemStack = tempInv.getStack(((ISlot) slots.get(i)).getIndex());
+                if (!itemStack.isEmpty() || override) { // same as if (!(itemStack.isEmpty() && !override))
```
Comment (Earthcomputer) [inline src/main/java/net/earthcomputer/clientcommands/command/KitCommand.java]: Unnecessary comment here

File: src/main/java/net/earthcomputer/clientcommands/command/KitCommand.java [hunk 21]
```diff
@@ -0,0 +1,191 @@
...
+            CompoundTag compoundTag = new CompoundTag();
+            kits.forEach(compoundTag::put);
+            rootTag.putInt("DataVersion", SharedConstants.getGameVersion().getWorldVersion());
+            rootTag.put("Kits", compoundTag);
+            File newFile = File.createTempFile("kits", ".dat", configPath.toFile());
+            NbtIo.write(rootTag, newFile);
+            File backupFile = new File(configPath.toFile(), "kits.dat_old");
+            File currentFile = new File(configPath.toFile(), "kits.dat");
+            Util.backupAndReplace(currentFile, newFile, backupFile);
+        } catch (Exception e) {
```
Comment (Earthcomputer) [inline src/main/java/net/earthcomputer/clientcommands/command/KitCommand.java]: Should probably catch `IOException` here.

File: src/main/java/net/earthcomputer/clientcommands/command/KitCommand.java [hunk 22]
```diff
@@ -0,0 +1,191 @@
...
+                    ListTag updatedListTag = new ListTag();
+                    compoundTag.getList(key, NbtType.COMPOUND).forEach(tag -> {
+                        Dynamic<Tag> oldTagDynamic = new Dynamic<>(NbtOps.INSTANCE, tag);
+                        Dynamic<Tag> newTagDynamic = client.getDataFixer().update(TypeReferences.ITEM_STACK, oldTagDynamic, rootTag.getInt("DataVersion"), version);
+                        updatedListTag.add(newTagDynamic.getValue());
+                    });
+                    kits.put(key, updatedListTag);
+                });
+            }
+        } catch (Exception e) {
```
Comment (Earthcomputer) [inline src/main/java/net/earthcomputer/clientcommands/command/KitCommand.java]: `IOException`

File: src/main/java/net/earthcomputer/clientcommands/command/KitCommand.java [hunk 23]
```diff
@@ -0,0 +1,191 @@
...
+    private static void loadFile() throws CommandSyntaxException {
+        try {
+            kits.clear();
+            CompoundTag rootTag = NbtIo.read(new File(configPath.toFile(), "kits.dat"));
+            if (rootTag == null) {
+                return;
+            }
+            final int version = SharedConstants.getGameVersion().getWorldVersion();
+            CompoundTag compoundTag = rootTag.getCompound("Kits");
+            if (rootTag.getInt("DataVersion") == version) {
```
Comment (Earthcomputer) [inline src/main/java/net/earthcomputer/clientcommands/command/KitCommand.java]: Could extract `rootTag.getInt("DataVersion")` to a variable. Also, this should be `>=` in case the kits file was saved in a future version.

**Review by Earthcomputer — CHANGES_REQUESTED** [review]
Almost there, just a few more things

File: src/main/java/net/earthcomputer/clientcommands/command/KitCommand.java (hunk 20 above)
Comment (xpple) [reply src/main/java/net/earthcomputer/clientcommands/command/KitCommand.java]: I chose to add that comment because `if (!itemStack.isEmpty() || override)` might seem totally arbritary to anyone who doesn't know what is being checked here: if not (the item stack is empty and override is not true) then... That makes more sense to read than some logical OR.

Lmk if you still find it unnecessary

Comment (Earthcomputer) [reply src/main/java/net/earthcomputer/clientcommands/command/KitCommand.java]: It's quite obvious, we all know de-morgan's law

Comment (xpple) [reply src/main/java/net/earthcomputer/clientcommands/command/KitCommand.java]: The what?! I have no idea what that is 🤣

File: src/main/java/net/earthcomputer/clientcommands/command/KitCommand.java (hunk 19 above)
Comment (xpple) [reply src/main/java/net/earthcomputer/clientcommands/command/KitCommand.java]: Should I remove that?

Comment (Earthcomputer) [reply src/main/java/net/earthcomputer/clientcommands/command/KitCommand.java]: You need one of them, check which one you need

File: src/main/java/net/earthcomputer/clientcommands/command/KitCommand.java [hunk 24]
```diff
@@ -74,9 +75,7 @@ public static void register(CommandDispatcher<ServerCommandSource> dispatcher) {
     }
 
     private static int create(ServerCommandSource source, String name) throws CommandSyntaxException {
-        loadFile();
```
Comment (Earthcomputer) [inline src/main/java/net/earthcomputer/clientcommands/command/KitCommand.java]: Humm, why did you delete this line?

**Comment by xpple** [comment]
Because I'm a fucking idiot

**Review by Earthcomputer — APPROVED** [review]
Tyvm!

**Comment by xpple** [comment]
Finallyyyyyyy thank you as well!
